### PR TITLE
resources: Add source code for riscv-tests

### PR DESCRIPTION
**Pull request by kunpai**
- This code is compatible with m5ops, i.e., ROIs are set using m5_work_begin and m5_work_end.
- Added README.md and CHANGELOG.md for every benchmark to track issues
- Updated base README.md of asmtest directory to show what code has been changed
- riscv-tests benchmarks' source code is now housed inside src/asmtest/benchmarks. Earlier, it used to be in src/riscv-tests

Change-Id: I17572d00b50617bc52e8453ee873b0c2f614f461

### CODE REVIEW

File: src/asmtest/.gitignore
Comment (powerjg) [inline src/asmtest/.gitignore:1]: Probably should be in a separate commit

File: src/asmtest/benchmarks/Makefile [hunk 1]
```diff
@@ -23,32 +23,44 @@ bmarks = \
...
 
 #--------------------------------------------------------------------
 # Build rules
 #--------------------------------------------------------------------
 
-RISCV_PREFIX ?= riscv$(XLEN)-unknown-elf-
+RISCV_PREFIX ?= riscv$(XLEN)-linux-gnu-
 RISCV_GCC ?= $(RISCV_PREFIX)gcc
-RISCV_GCC_OPTS ?= -DPREALLOCATE=1 -mcmodel=medany -static -std=gnu99 -O2 -ffast-math -fno-common -fno-builtin-printf -fno-tree-loop-distribute-patterns
+RISCV_GCC_OPTS ?= -DPREALLOCATE=1 -mcmodel=medany -static -std=gnu99 -O2 -ffast-math -fno-common -fno-builtin-printf -fno-tree-loop-distribute-patterns -march=rv$(XLEN)gcv -mabi=lp64d
```
Comment (powerjg) [inline src/asmtest/benchmarks/Makefile:46]: Not sure it  makes sense to allow `XLEN` to change and not update `mabi`

File: src/asmtest/benchmarks/Makefile [hunk 2]
```diff
@@ -23,32 +23,44 @@ bmarks = \
...
 RISCV_LINK ?= $(RISCV_GCC) -T $(src_dir)/common/test.ld $(incs)
 RISCV_LINK_OPTS ?= -static -nostdlib -nostartfiles -lm -lgcc -T $(src_dir)/common/test.ld
 RISCV_OBJDUMP ?= $(RISCV_PREFIX)objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data
-RISCV_SIM ?= spike --isa=rv$(XLEN)gc
+RISCV_SIM ?= spike
```
Comment (powerjg) [inline src/asmtest/benchmarks/Makefile:50]: Why this change?

File: src/asmtest/benchmarks/Makefile [hunk 3]
```diff
@@ -23,32 +23,44 @@ bmarks = \
...
+# old compile template
+# define compile_template
+# $(1).riscv: $(wildcard $(src_dir)/$(1)/*) $(wildcard $(src_dir)/common/*)
+# 	$$(RISCV_GCC) $$(incs) $$(RISCV_GCC_OPTS) -o $$@ $(wildcard $(src_dir)/$(1)/*.c) $(wildcard $(src_dir)/$(1)/*.S) $(wildcard $(src_dir)/common/*.c) $(wildcard $(src_dir)/common/*.S) $$(RISCV_LINK_OPTS)
+# endef
+
 define compile_template
 $(1).riscv: $(wildcard $(src_dir)/$(1)/*) $(wildcard $(src_dir)/common/*)
-	$$(RISCV_GCC) $$(incs) $$(RISCV_GCC_OPTS) -o $$@ $(wildcard $(src_dir)/$(1)/*.c) $(wildcard $(src_dir)/common/*.c) $(wildcard $(src_dir)/common/*.S) $$(RISCV_LINK_OPTS)
+	$$(RISCV_GCC) -I$(src_dir)/common -static -O2 -o $$@ $(wildcard $(src_dir)/$(1)/*.c) $(wildcard $(src_dir)/$(1)/*.S) $(src_dir)/common/gem5/util/m5/build/riscv/out/libm5.a
```
Comment (powerjg) [inline src/asmtest/benchmarks/Makefile:63]: This makes an assumption about where gem5/ is. You should make sure to document this or make a variable.

File: src/asmtest/benchmarks/common/crt.S [hunk 4]
```diff
@@ -47,8 +47,8 @@ _start:
   li  x30,0
   li  x31,0
 
-  # enable FPU and accelerator if present
-  li t0, MSTATUS_FS | MSTATUS_XS
+  # enable FPU, vector, and accelerator if present
+  li t0, MSTATUS_FS | MSTATUS_XS | MSTATUS_VS
```
Comment (powerjg) [inline src/asmtest/benchmarks/common/crt.S:51]: Needed?

File: src/asmtest/benchmarks/common/encoding.h
Comment (powerjg) [inline src/asmtest/benchmarks/common/encoding.h:1]: Can you explain these changes?

**Review by powerjg — COMMENTED** [review]
A few small comments after going through the first few files.

Overall.. can you explain what you wrote and what came directly from the upstream?

File: src/asmtest/benchmarks/Makefile (hunk 1 above)
Comment (kunpai) [reply src/asmtest/benchmarks/Makefile:46]: This change came from riscv-tests. Should I update mabi to use `lp($XLEN)d` instead?

**Comment by kunpai** [comment]
> A few small comments after going through the first few files.
> 
> Overall.. can you explain what you wrote and what came directly from the upstream?

Mostly, I changed the compile_template of the Makefile to not use the riscv-tests linker. In the source code, I did not change much, except comment out places where spike magic instructions are being called (and it is not important to the run of the program, Start_Timer(); in dhrystone for example) and replace their setStats calls with m5ops. I have documented what I changed in the README.md of src/asmtest.

File: src/asmtest/benchmarks/common/encoding.h
Comment (kunpai) [reply src/asmtest/benchmarks/common/encoding.h:1]: Since I did not use the riscv-tests linkers and just compiled with linking to the `src/asmtest/benchmarks/common/` directory and m5ops, they could not find `encoding.h`, which was in a different directory. So, I just put `encoding.h` into the `common` directory. I found trouble using m5ops if I linked with two different directories while compiling.

File: src/asmtest/benchmarks/common/crt.S (hunk 4 above)
Comment (kunpai) [reply src/asmtest/benchmarks/common/crt.S:51]: I am not really sure. This came from `riscv-tests`.

File: src/asmtest/benchmarks/Makefile (hunk 3 above)
Comment (kunpai) [reply src/asmtest/benchmarks/Makefile:63]: I have documented this in the README.md of `src/asmtest`

File: src/asmtest/benchmarks/Makefile (hunk 2 above)
Comment (kunpai) [reply src/asmtest/benchmarks/Makefile:50]: I was playing around with different compilation options and I forgot to reset it. I will add this back. Thanks for pointing this out!

**Comment by powerjg** [comment]
It would be great if you could separate out your changes from the changes
that were introduced upstream.

The best would be to have 3 commits:
- Moving the files around and deleting unneeded files
- Updating to latest riscv-tests upstream
- Your changes.

If this isn't possible, then we can figure out how to review it as-is.

**Comment by kunpai** [comment]
> It would be great if you could separate out your changes from the changes that were introduced upstream. The best would be to have 3 commits: - Moving the files around and deleting unneeded files - Updating to latest riscv-tests upstream - Your changes. If this isn't possible, then we can figure out how to review it as-is.
> […](#)

Okay, sounds good to me. Let me see what I can do. For now, I will close this pull request.
